mayastor/src/bdev/nexus/nexus_persistence.rs

```diff
@@ -0,0 +1,146 @@
+use crate::{
+    bdev::{
+        nexus::nexus_child::ChildState::Faulted,
+        ChildState,
+        Nexus,
+        Reason,
+    },
+    persistent_store::PersistentStore,
+    sleep::mayastor_sleep,
+};
+use rpc::persistence::{
+    ChildInfo,
+    ChildState as PersistentChildState,
+    NexusInfo,
+    Reason as PersistentReason,
+};
+use std::time::Duration;
+
+type ChildUuid = String;
+
+/// Defines the type of persist operations.
+pub(crate) enum PersistOp {
+    /// Create a persistent entry.
+    Create,
+    /// Update a persistent entry.
+    Update((ChildUuid, ChildState)),
+    /// Save the clean shutdown variable.
+    Shutdown,
+}
+
+impl Nexus {
+    /// Persist information to the store.
+    pub(crate) async fn persist(&self, op: PersistOp) {
+        if !PersistentStore::enabled() {
+            return;
+        }
+        let mut nexus_info = self.nexus_info.lock().await;
+        match op {
+            PersistOp::Create => {
+                // Initialisation of the persistent info will overwrite any
+                // existing entries.
+                // This should only be called on nexus creation.
+                nexus_info.clean_shutdown = false;
+                self.children.iter().for_each(|c| {
```

Comment:
Shouldn't we also reset 'nexus_info.children' array to make sure it's clear and contains no stalled information when we repopulate it ?

Reply:
This code is only ever called when we are creating the nexus. As a result, the expectation is that the NexusInfo structure has only just been created and cannot contain anything yet. I will make this clearer by including an `assert` which checks that the vector of children is empty.

Reply:
I've added a couple of asserts here to check the values of the data structure.